File: CivitasJava/civitas/JugadorEspeculador.java
package civitas;

import java.util.ArrayList;

public class JugadorEspeculador extends Jugador {
    
    private static int FactorEspeculador = 2;
    private float fianza;
    
    protected JugadorEspeculador(Jugador otro, float fianza){
        super(otro);
        this.fianza = fianza;
        
        for(int i = 0; i < propiedades.size(); i++){
            propiedades.get(i).actualizaPropietarioPorConversion(this);
        }
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel){
        if ( debeSerEncarcelado() ){
            if (puedoGastar(fianza) ){
                paga(fianza);
                Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre() + " se libra de la cárcel pagando fianza");
                
            }else{
                moverACasilla(numCasillaCarcel);
                encarcelado = true;
                Diario.getInstance().ocurreEvento("El jugador especulador " + getNombre() + " ha sido encarcelado");
            }
        }
        
        return encarcelado;
    }
    
    @Override
    boolean pagaImpuesto (float cantidad){
        if (isEncarcelado() )
            return false;
        else
            return paga(cantidad / 2);
    }
    
    @Override
    int getCasasMax(){
        return (CasasMax * FactorEspeculador);
    }
    
    @Override
    int getHotelesMax(){
        return (HotelesMax * FactorEspeculador);
    }
    
    @Override
    public String toString(){
       return ("Jugador especulador " + getNombre() + ", saldo " + getSaldo() + ", está en la casilla " + getNumCasillaActual() + "\n"); 
    }
}

File: CivitasRuby/jugador_especulador.rb
require_relative 'jugador'

module Civitas
class JugadorEspeculador < Jugador

  def initialize(otro_jugador, fianza)
    super(otro_jugador.nombre)
    self.copia(otro_jugador)
    @@FactorEspeculador = 2
    @fianza = fianza
  end
  
  def encarcelar(num_casilla_carcel)
    if debe_ser_encarcelado
      if puedo_gastar(@fianza)
        paga(@fianza)
        Diario.instance.ocurre_evento("El jugador especulador " + @nombre + " se libra de la cárcel pagando fianza de " + @fianza)
      else
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("El jugador " + @nombre + " ha sido encarcelado")
      end
    end
  end
  
  def paga_impuesto(cantidad)
    if @encarcelado
      false;
    else
      self.paga(cantidad / 2);
    end
  end
  
  def to_string
    ("Jugador especulador " + @nombre + ", saldo " + @saldo + ", está en la casilla " + @numCasillaActual)
  end
end
end
